### DISCUSSION TITLE
p.export() with "download_missing"

### DISCUSSION THREAD

**Discussion by moffat** [RhetTbull/osxphotos#624] [Q&A] [answered]
Hi,

I'm trying go write a script that exports all shared albums to  `backup/SHAREDALBUMNAME` folder. I think something like this should work:

```python
import osxphotos
import os 

photosdb = osxphotos.PhotosDB()
albums = photosdb.albums_shared

basedir = '~/Local/backups/shared_albums'



for a in albums:
    edir = os.path.join(basedir,a)
    for p in photosdb.photos(albums=[a]):
        p.export(edir)
```

(the code assumes the folders already exist)

Now, my problem is that with shared albums, many files are not downloaded, so I need to pass the `download_missing` options. But `p.export(edit,download_missing=True) returns

```python
TypeError: export() got an unexpected keyword argument 'download_missing'
```

and indeed help(p.export()) doesn't list "download_missing". What am I... missing? 

Thanks.

**Comment by RhetTbull** [comment, marked as answer]
Yeah, the export() interface exposed by PhotoInfo is pretty limited and meant for quick export but lacks a lot of options like download_missing. You could use "use_photos_export=True" and that will tell export to always use the interface to Photos to export the file (and download it if necessary).  Perhaps I should add "download_missing" but there's a more robust [PhotoExporter](https://github.com/RhetTbull/osxphotos#photoexporter) interface that exposes a lot more capability.  

try this (I added a few lines so I could actually run the code to test):

```python
import osxphotos
import os 
from osxphotos.photoexporter import ExportOptions, PhotoExporter

photosdb = osxphotos.PhotosDB()
albums = photosdb.albums_shared

basedir = '~/Local/backups/shared_albums'

for a in albums:
    edir = os.path.expanduser(os.path.join(basedir,a))
    os.makedirs(edir,exist_ok=True)
    for p in photosdb.photos(albums=[a]):
        PhotoExporter(p).export(edir, options=ExportOptions(download_missing=True))
```

**Reply by RhetTbull** [reply]
Note that export can get complicated quickly...do you want the edited version or original or both? Do you want the live photo video file? Do you also want the RAW complement for RAW+JPEG?  etc.  `ExportOptions` has options for dealing with these things.  One thing to note is that you can currently download either the original version or the edited version but not both in one call to `export()` so you'll need to call `export()` a second time if you want to also export the edited version. e.g. 

```python
for a in albums:
    edir = os.path.expanduser(os.path.join(basedir,a))
    os.makedirs(edir,exist_ok=True)
    for p in photosdb.photos(albums=[a]):
        PhotoExporter(p).export(edir, options=ExportOptions(download_missing=True))
        if p.hasadjustments:
             PhotoExporter(p).export(edir, filename=edited_filename, options=ExportOptions(edited=True, download_missing=True))
```

**Reply by moffat** [reply]
Ah, thanks a lot for the explanation, that makes sense. Your modified code worked great!

**Reply by moffat** [reply]
I did notice some of the albums didn't download. I tried it with one of those albums with osxphotos instead and got:

```bash

  [####################################]  1543/1543
Processed: 1543 photos, exported: 0, missing: 0, error: 0
Elapsed time: 0:04:08
cygnus:bin$ osxphotos -v
osxphotos, version 0.45.4

```

this is in Monterey.

**Reply by RhetTbull** [reply]
Run your command with the `--verbose` flag which will print out a while bunch of output.  Look for a line like this for one of the files that doesn't get exported:

`Exporting IMG_1997.JPG (4D521201-92AC-43E5-8F7C-59BC41C37A96.jpeg) as IMG_1997.JPG (16/27)`

The unique ID (UUID) of the file is the part in parentheses that looks like this: `4D521201-92AC-43E5-8F7C-59BC41C37A96`

Copy that UUID then run the following command (substituting the UUID you copied)

`osxphotos debug-dump --dump photos --uuid 4D521201-92AC-43E5-8F7C-59BC41C37A96 > debug.txt`

Post the debug.txt file here and I'll see if I can find any clues.  Note: the file may contain location information so if that's sensitive and you want to sanitize, do so before posting.  (look for `latitude` and `longitude`)

I don't have a machine capable of running Monterey so I've had to do all testing "remotely" like this...sorry.

**Reply by moffat** [reply]
See attached. [debug.txt](https://github.com/RhetTbull/osxphotos/files/8008952/debug.txt)

Thanks!

**Reply by RhetTbull** [reply]
Thanks for the debug data. Nothing jumps out as a smoking gun -- looks like a normal shared photo.  What happens if you try to export a photo from the album in question using Photos? (Using the `File | Export` menu)  The specific photo in question is named "IMG_1277.JPG" so you might try searching for that to find that one.  

I've been trying to replicate this but right now my Photos library won't export any shared photos (even from Photos itself) so I'll try repairing the library.  

No matter what the issue is, osxphotos should give more detail (e.g. an error or explanation for why the photo was skipped) so I'll have to fix that once I get my library working.

**Reply by moffat** [reply]
I tried this and the export worked fine, as far as I can tell. 

Interestingly, after I tried doing the manual export as you suggested, and trying osxphotos again, the entire album exported correctly.

**Reply by RhetTbull** [reply]
I've not been able to replicate this. Will try to figure out where to add some extra reporting in the code to make it more clear where the issue is.  I just repaired my Photos library and shared photos are correctly being exported.  (I'm not using Monterey though and I have had some reports (#545) of problems with shared photos on Monterey)

In the meantime, I suggest you try the osxphotos command you tried earlier but also add the `--use-photokit` flag in addition to `--download-missing`.  Does that give a different outcome?   If that doesn't work, one option is to try repairing your Photos library (hold down `option+command` when starting Photos) then try again.  I've noticed that repairing the library can sometimes fix problems with shared photos.

**Reply by moffat** [reply]
Ok, so the behaviour is more inconsistent that I thought. I did try to repair my library *before* I read your last message, as I had read that could sort issues with regular export of photos.

I tried downloading a different album where my code had resulted in an empty folder. This time *some* of the photos/videos were downloaded (maybe 10 out of 125). I tried again, and the count went up to ~30. 

As for your second suggestion, using `--use-photokit` fails with: 

```bash
Error exporting photo (CED06217-BEC1-47CD-A539-3E36017C24FC: IMG_0107.JPG) as IMG_0107.JPG: Could not get authorizaton to use Photos: auth_status = 0
```

Add: python actually crashes when using `--use-photokit`. log attached. 
[osxphotos_crash.txt](https://github.com/RhetTbull/osxphotos/files/8017539/osxphotos_crash.txt)

**Reply by RhetTbull** [reply]
I created new issues for both these (#626 , #625 )

**Reply by moffat** [reply]
Great, thanks!

**Comment by RhetTbull** [comment]
> Add: python actually crashes when using --use-photokit. log attached.

Ugh.  The Terminal is supposed to request the authorization (that's what happens on Catalina and Big Sur).  I won't be able to fix this until I get access to a machine with Monterey (and even then, it might be impossible to fix).  

>Ok, so the behaviour is more inconsistent that I thought. 

When I get a chance I'll try to add some more instrumentation in so you can collect debugging info.  Not really sure what the issue is.

**Reply by moffat** [reply]
@RhetTbull thank you, I appreciate all your feedback. I was able to backup some of my shared albums so this is very helpful.
